### PR TITLE
Add profiling to training engine

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

This PR introduces step-driven profiling to the MaxText training engine to enable detailed performance tracing using JAX's profiling tools.

Specifically, this change:
- **Introduces a new `Profiler` class** (`src/maxtext/training_engine/profiler.py`) that manages the opening and closing of a profiling window based on the training configuration. 
- **Adds a `@_profiled_step` decorator** in `maxtext_engine.py` to automatically wrap execution steps (such as `fwd_bwd` and `update`) with `jax.profiler.StepTraceAnnotation`. This ensures regions are properly tagged with their `step_num` in the trace viewer.
- **Automatically drives the profiler** by activating it on the first call of an opened profile and deactivating it on the closing step.
- Ensures that any active profile is safely stopped if the engine shuts down unexpectedly via `engine.close()`.

# Tests

CI tests
Profile for Qwen3-0.6B local run: https://xprof.corp.google.com/trace_viewer/sjsurbhi-705674918985334607

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/training_engine/micro_step_profiler.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Copyright 2023-2026 Google LLC
```
Comment (igorts-git) [inline src/maxtext/training_engine/micro_step_profiler.py:1]: I recommend renaming this file to `micro_step_profiler.py`, because Maxtext already has `profiler.py`

**Review by igorts-git — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]

File: src/maxtext/training_engine/micro_step_profiler.py (hunk 1 above)
Comment (SurbhiJainUSC) [reply src/maxtext/training_engine/micro_step_profiler.py:1]: sounds good.
